### PR TITLE
GeofenceRegion.Payload

### PR DESCRIPTION
**Pull request by dahlbyk**
### Description of Change ###

Adds optional `GeofenceRegion.Payload` for saving other useful stuff, e.g. a display name to use in notifications.

~~Also adjusts `Position` to be more serializable, so the `Payload` change can be tested. I'm not sure why this adjustment seems to be necessary, since saving geofences clearly works?~~ Nevermind, I found the `GeofenceRegionStore` wrapper for flattening the data structure for storage.

### Issues Resolved ### 
N/A

#### Shiny.Locations.GeofenceRegion
- `Dictionary<string, object>? Payload { get; set; }`

### Platforms Affected ### 

- All

### Behavioral Changes ###

None

### Testing Procedure ###

Unit tests for serializing/deserializing `GeofenceRegionStore` with `Payload`

### PR Checklist ###

- [x] Rebased on top of the target branch at time of PR
    - PR targets `master` because `dev` build is broken
- [ ] Changes adhere to coding standard

### CODE REVIEW

**Comment by dahlbyk** [comment]
Just thinking out loud - the serialization tests would be cleaner if they round-tripped `GeofenceRegion` instead of `GeofenceRegionStore`, but the conversion logic is buried in `Wrap()`.

**Comment by dahlbyk** [comment]
Rebased onto `master`. I assume org owner has permission to retarget PR to `dev` if you prefer?

**Comment by aritchie** [comment]
I moved it to dev.

I need some time to think about additions like this.  Notifications only has it because iOS has it natively and because Android requires a unique INT ID which isn't as easy to correlate.  Jobs also has this because it really has to.  Geofences though, have a good identifier though that can be traced to something in a user database.  

Let me think about this one.  There is value, just not sure if I want to start ballooning the args yet

File: src/Shiny.Locations/LocationExtensions.cs [hunk 1]
```diff
@@ -36,7 +36,8 @@ public static async Task<AccessState> RequestAccessAndStart(this IGpsManager gps
                 RadiusMeters = args.Radius.TotalMeters,
                 SingleUse = args.SingleUse,
                 NotifyOnEntry = args.NotifyOnEntry,
-                NotifyOnExit = args.NotifyOnExit
+                NotifyOnExit = args.NotifyOnExit,
+                Payload = args.Payload.Cast<GeofenceRegionStore.PayloadEntry>().ToList(),
```
Comment (dahlbyk) [inline src/Shiny.Locations/LocationExtensions.cs]:
```suggestion
                Payload = args.Payload?.Cast<GeofenceRegionStore.PayloadEntry>().ToList(),
```

File: src/Shiny.Locations/LocationExtensions.cs [hunk 2]
```diff
@@ -46,7 +47,8 @@ public static async Task<AccessState> RequestAccessAndStart(this IGpsManager gps
             {
                 SingleUse = store.SingleUse,
                 NotifyOnEntry = store.NotifyOnEntry,
-                NotifyOnExit = store.NotifyOnExit
+                NotifyOnExit = store.NotifyOnExit,
+                Payload = store.Payload.ToDictionary(kvp => kvp.Key, kvp => kvp.Value),
```
Comment (dahlbyk) [inline src/Shiny.Locations/LocationExtensions.cs]:
```suggestion
                Payload = store.Payload?.ToDictionary(kvp => kvp.Key, kvp => kvp.Value),
```

**Comment by dahlbyk** [comment]
> need to test serialization without `Payload`.

Fixed. Of course `XmlSerializer` can't represent serializing `null`, so it deserializes an empty list. 🙄

**Comment by aritchie** [comment]
@dahlbyk Appreciate the PR as always.  This one actually isn't necessary since delegates have a built in mechanism for state.  If you make a MVVM "viewmodelish" property with a get/set in the delegate, that state is serialized/deserialize to settings.

**Comment by dahlbyk** [comment]
> This one actually isn't necessary since delegates have a built in mechanism for state.

Sounds promising. 😀 

> If you make a MVVM "viewmodelish" property with a get/set in the delegate...

By "viewmodelish" do you mean implementing INPC (or `[Reactive]` or whatever), or just a plain property on the `IGeofenceDelegate` implementation?

> that state is serialized/deserialize to settings.

Pardon my ignorance, but I'm trying to figure out what the lifecycle looks like on this. If I resolve the current delegate and update its state, what causes that value to persist through to the next app boot? Or am I thinking about the state update wrong?

**Comment by dahlbyk** [comment]
> > If you make a MVVM "viewmodelish" property with a get/set in the delegate...
> 
> By "viewmodelish" do you mean implementing INPC (or `[Reactive]` or whatever), or just a plain property on the `IGeofenceDelegate` implementation?

Dug a bit and I'm guessing that's exactly what you mean, referring to this:

https://github.com/shinyorg/shiny/blob/af46ecfee901d8d5edc50ee3cee2a739f0ecef99/src/Shiny.Core/Infrastructure/ShinyServiceCollection.cs#L47-L72

**Comment by aritchie** [comment]
That's the one :)

**Pull request closed without merging**